### PR TITLE
.github, README: sign release archives with minisign

### LINKED ISSUE #548: build: sign releases
As of commit https://github.com/exercism/configlet/commit/f7c12253077d1ed3b70f4453fb30e0c8db1121de, each configlet release has a checksums file. Let's sign releases too.

I'd suggest using `minisign`, which is a portable and compatible implementation of OpenBSD's `signify` protocol. 

Background:

- https://latacora.micro.blog/2019/07/16/the-pgp-problem.html (Ctrl+f minisign)
- https://flak.tedunangst.com/post/signify
- https://www.openbsd.org/papers/bsdcan-signify.html
- https://wiki.debian.org/Teams/Apt/Spec/AptSign (Debian moving from PGP to Ed25519)
- https://buttondown.email/cryptography-dispatches/archive/cryptography-dispatches-hello-world-and-openpgp/  (from Filippo Valsorda, who is in charge of cryptography/security for Go).
- https://buttondown.email/cryptography-dispatches/archive/cryptography-dispatches-replace-pgp-with-an-https/
- https://blog.gtank.cc/modern-alternatives-to-pgp/
- https://docs.sigstore.dev/rekor/sign-upload/

`minisign` info:

- https://github.com/jedisct1/minisign
- https://jedisct1.github.io/minisign/
- https://man.archlinux.org/man/community/minisign/minisign.1.en

`minisign` installation:

- https://github.com/jedisct1/minisign/releases (has binaries for Linux/macOS/Windows)
- https://archlinux.org/packages/community/x86_64/minisign/
- https://src.fedoraproject.org/rpms/minisign
- https://formulae.brew.sh/formula/minisign

Some projects that sign releases using `minisign` or `signify`:

- [minisign](https://github.com/jedisct1/minisign/releases) itself
- [libsodium](https://github.com/jedisct1/libsodium/releases)
- [OpenBSD](https://cdn.openbsd.org/pub/OpenBSD/7.0/amd64/)
- [Void Linux](https://docs.voidlinux.org/installation/index.html#verifying-images)
- [OpenWRT](https://downloads.openwrt.org/releases/21.02.2/packages/aarch64_generic/base/Packages.sig)
- [GrapheneOS](https://grapheneos.org/install/cli#obtaining-factory-images)
- [WireGuard for Android kernel modules](https://git.zx2c4.com/wireguard-android/)
- [LibreSSL](https://www.libressl.org/releases.html)
- [LibreSSL portable](https://github.com/libressl-portable/portable/blob/b52dc3d9b292f4f644d7506a2d62df11f2a6e269/libressl.pub) - public key in git repo
- [dnscrypt-proxy](https://github.com/DNSCrypt/dnscrypt-proxy/releases) (with instructions [here](https://github.com/DNSCrypt/dnscrypt-proxy/wiki/Installation#verification-of-downloaded-files))
- [Zig](https://ziglang.org/download/) (see also [this issue](https://www.github.com/ziglang/zig/issues/4945))

`signify` info:

- https://www.openbsd.org/faq/faq4.html#Download
- https://man.openbsd.org/signify

`signify` installation:

- Arch Linux: https://archlinux.org/packages/community/x86_64/signify/
- Debian: https://packages.debian.org/stable/main/signify-openbsd
- Ubuntu: https://packages.ubuntu.com/focal/signify-openbsd
- https://github.com/aperezdc/signify portable version

---

To verify a file with `minisign`:

```shell
minisign -V -m SOURCE_FILE -P PUBLIC_KEY
```

if the signature file `SOURCE_FILE.minisig` is present in the same directory.

Otherwise:

```shell
minisign -V -m SOURCE_FILE -P PUBLIC_KEY -x SIGNATURE_FILE
```

`minisign` can verify `signify` signatures, but `signify` cannot verify `minisign` signatures (because `minisign` adds an extra "trusted comment" line).

> The main addition is a "trusted comment" line, that can be used to verify metadata, instead of just the content, for example to verify a timestamp and prevent unwanted downgrades. This is not an issue with packages, but it can be an issue with files that don't contain any version/timestamp.

> Signature files include an untrusted comment line that can be freely modified, even after signature creation.
>
> They also include a second comment line, that cannot be modified without the secret key.
>
> Trusted comments can be used to add instructions or application-specific metadata (intended file name, timestamps, resource identifiers, version numbers to prevent downgrade attacks).

> Trusted comments are important to describe what has been signed in addition to the fact that something has been signed.

See also: [jedisct1/minisign/#59](https://github-redirect.dependabot.com/jedisct1/minisign/issues/59)

---

One alternative is signing with SSH keys.

Links:

- https://www.agwa.name/blog/post/ssh_signatures

### PR DESCRIPTION
**Pull request by ee7**
Since commit f7c12253077d, each configlet release has a checksums file. The assets for the [most recent release](https://github.com/exercism/configlet/releases/tag/4.0.0-beta.12) were:

```text
configlet_4.0.0-beta.12_checksums_sha256.txt
configlet_4.0.0-beta.12_linux_x86-64.tar.gz
configlet_4.0.0-beta.12_macos_x86-64.tar.gz
configlet_4.0.0-beta.12_windows_x86-64.zip
```

But confirming a checksum match only checks for accidental corruption. Add signing with [minisign][1] to the release process, such that the next release will have a `.minisig` file for each archive:

```text
configlet_4.0.0-beta.13_checksums_sha256.txt
configlet_4.0.0-beta.13_linux_x86-64.tar.gz
configlet_4.0.0-beta.13_linux_x86-64.tar.gz.minisig
configlet_4.0.0-beta.13_macos_x86-64.tar.gz
configlet_4.0.0-beta.13_macos_x86-64.tar.gz.minisig
configlet_4.0.0-beta.13_windows_x86-64.zip
configlet_4.0.0-beta.13_windows_x86-64.zip.minisig
```

From the [minisign docs][2]:

> Minisign is a dead simple tool to sign files and verify signatures.
> It is portable, lightweight, and uses the highly secure Ed25519 public-key signature system.

For example, to verify the above Linux x86-64 release archive, run:

```shell
minisign -Vm configlet_4.0.0-beta.13_linux_x86-64.tar.gz -P RWQGj6DTXgYLhKvWJMGtbDUrZerawUcyWnti9MGuWMx7VDW9DqZn2tMZ
```

where the argument to -P is the configlet minisign public key. Alternatively, you can download the `configlet-minisign.pub` file and run:

```shell
minisign -Vm configlet_4.0.0-beta.13_linux_x86-64.tar.gz -p configlet-minisign.pub
```

Minisign is an implementation of OpenBSD's signify protocol. For more background, see posts on [signify][3] and [problems with PGP][4].

An alternative would be to [sign releases with an SSH key][5], but minisign is more focused, always uses Ed25519, and is simpler. Especially for verification.

Some other projects that sign releases with minisign or signify:

- [libsodium][6]
- [OpenBSD][7]
- [Void Linux][8]
- [Zig][9]
- [LibreSSL][10]

For now, the fetch-configlet scripts don't support verifying signatures. They may be able to do so eventually, but it won't be required: we don't want to require the user to have minisign installed.

Don't upload a signature for the checksum file, because:

- We want to optimize for the simplicity of verifying a single configlet release archive, so it's best to have only one way to do it: verify the signature for that archive.
- It's harder to explain the alternative: verify the signature for the checksum file AND check that the sha256 of the release archive matches that in the checksum file.
- It doesn't matter that verifying `n` configlet release archives requires running `n` minisign commands, rather than exactly 1 minisign command plus 1 `sha256sum --check` (or `sha256 -c` on some platforms) command.

[1]: https://github.com/jedisct1/minisign
[2]: https://jedisct1.github.io/minisign/
[3]: https://www.openbsd.org/papers/bsdcan-signify.html
[4]: https://latacora.micro.blog/2019/07/16/the-pgp-problem.html
[5]: https://www.agwa.name/blog/post/ssh_signatures
[6]: https://github.com/jedisct1/libsodium/releases/tag/1.0.18-RELEASE
[7]: https://www.openbsd.org/faq/faq4.html#Download
[8]: https://docs.voidlinux.org/installation/index.html#verifying-digital-signature
[9]: https://github.com/ziglang/zig/releases/tag/0.11.0
[10]: https://www.libressl.org/releases.html

See #548 for more background.

Closes: #548

---

### Testing

I've tested that this PR works on a [test release in the exercism/configlet repo](https://github.com/exercism/configlet/releases/tag/untagged-da7afb751e345a2c46ee).

Verifying the release:

```console
$ gh -R exercism/configlet release download 4.0.0-beta.13+sign.7 -D /tmp/configlet-sign
$ cd /tmp/configlet-sign
$ ls -1
configlet_4.0.0-beta.13+sign.7_checksums_sha256.txt
configlet_4.0.0-beta.13+sign.7_linux_x86-64.tar.gz
configlet_4.0.0-beta.13+sign.7_linux_x86-64.tar.gz.minisig
configlet_4.0.0-beta.13+sign.7_macos_x86-64.tar.gz
configlet_4.0.0-beta.13+sign.7_macos_x86-64.tar.gz.minisig
configlet_4.0.0-beta.13+sign.7_windows_x86-64.zip
configlet_4.0.0-beta.13+sign.7_windows_x86-64.zip.minisig
$ cat configlet_4.0.0-beta.13+sign.7_linux_x86-64.tar.gz.minisig
untrusted comment: signature from minisign secret key
RUQGj6DTXgYLhDchGz8qUXMfBPnIARrSv9Wc9bCD/oIS4/rnVa5o2CuAcWvcmvgqXRbzv9dYedV9fuicSdF2GqglZAzH4j95hQI=
trusted comment: timestamp:2023-07-18T11:07:19Z  file:configlet_4.0.0-beta.13+sign.7_linux_x86-64.tar.gz  hashed
9dsAfBpl5hx3egwKWEs0yIJMtqLAO7t/N+aRVEWLe9iGwp6nm7ChJTbl54pUPgB0UbY2BpEUICl1wFm4m3blDg==
$ minisign -Vm configlet_4.0.0-beta.13+sign.7_linux_x86-64.tar.gz -P RWQGj6DTXgYLhKvWJMGtbDUrZerawUcyWnti9MGuWMx7VDW9DqZn2tMZ
Signature and comment signature verified
Trusted comment: timestamp:2023-07-18T11:07:19Z  file:configlet_4.0.0-beta.13+sign.7_linux_x86-64.tar.gz  hashed
```

To-do:

- [x] Consider using `-W` (https://github.com/jedisct1/minisign/releases/tag/0.11)
- [x] Upload a separate `minisig` file for each asset
- [x] Consider not uploading the checksum minisig file
- [x] README: add initial documentation for verifying a configlet release archive
- [x] Tweak trusted comment format
- [x] Restrict permissions of `GITHUB_TOKEN`
- [x] Consider writing the public key in a `minisign.pub` file in this repo, and read from that file in the signing workflow.
- [x] Set the value of `CONFIGLET_MINISIGN_SECRET_KEY` as a secret in this repo (/environment/organization)
- [x] Create a test release with signed assets in this repo

### CODE REVIEW

File: .github/bin/upload-checksums-file [hunk 1]
```diff
@@ -13,5 +13,34 @@ cd "${download_dir}" || exit
...
 
-# Upload checksums file
-gh release upload "${build_tag}" "${checksums_file}"
+# Write secret key to expected location
+if [ -n "${MINISIGN_SECRET_KEY}" ]; then
+  minisign_dir="${HOME}/.minisign"
+  mkdir -p "${minisign_dir}"
+  printenv MINISIGN_SECRET_KEY > "${minisign_dir}/minisign.key"
+else
+  printf "MINISIGN_SECRET_KEY environmental variable not found. Exiting."
```
Comment (ErikSchierboom) [inline .github/bin/upload-checksums-file]:
```suggestion
  printf "MINISIGN_SECRET_KEY environment variable not found. Exiting."
```

File: .github/bin/upload-checksums-file [hunk 2]
```diff
@@ -13,5 +13,34 @@ cd "${download_dir}" || exit
...
+  exit 1
+fi
+
+# Write password (see https://github.com/jedisct1/minisign/issues/43)
+password_path='minisign_password.txt'
+if [ -n "${MINISIGN_PASSWORD}" ]; then
+  printenv MINISIGN_PASSWORD > "${password_path}"
+else
+  printf "MINISIGN_PASSWORD environmental variable not found. Exiting."
```
Comment (ErikSchierboom) [inline .github/bin/upload-checksums-file]:
```suggestion
  printf "MINISIGN_PASSWORD environment variable not found. Exiting."
```

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -70,8 +70,13 @@ jobs:
...
+      - name: Install minisign
+        run: ./.github/bin/linux-install-minisign
+
       - name: Upload checksums file
         run: ./.github/bin/upload-checksums-file
         env:
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
           REF: ${{ github.ref }}
+          MINISIGN_SECRET_KEY: ${{ secrets.MINISIGN_SECRET_KEY }}
+          MINISIGN_PASSWORD: ${{ secrets.MINISIGN_PASSWORD }}
```
Comment (ErikSchierboom) [inline .github/workflows/build.yml]: Are there any restrictions we need to know about when setting these secrets?

File: .github/bin/upload-checksums-file (hunk 1 above)
Comment (ee7) [reply .github/bin/upload-checksums-file]: Thanks. Pushed bcd4f04f82ea.

File: .github/bin/upload-checksums-file [hunk 4]
```diff
@@ -13,5 +13,34 @@ cd "${download_dir}" || exit
...
+fi
+
+# Sign checksum file
+dt="$(date -u '+%Y-%m-%dT%H:%M:%SZ')" # Like `date --utc --iso=seconds`
+# The below is the same format as minisign's default, but using a non-unix timestamp.
+trusted_comment="timestamp:${dt}   file:${checksums_file}    hashed"
+minisign -Sm "${checksums_file}" -t "${trusted_comment}" < "${password_path}"
+
+# Verify the signed file
+minisign_public_key='RWSDBrBOL29nTxsWe47s5pdnGLNvklqoM84yKr2oYvwhLMDVkpA51RnH'
```
Comment (ee7) [inline .github/bin/upload-checksums-file]: It's probably better to add a `minisign.pub` file to this repo, and then read from that file here.

We can also think about publishing it somewhere else.

File: .github/workflows/build.yml (hunk 3 above)
Comment (ee7) [reply .github/workflows/build.yml]: (Reproducing a long-lost Slack message).

Yes. But just:

- We generate the public/secret key pair using `minisign`
- We make the public key generally known - ideally mentioning it outside this repo too.
- The private key that we write here must correspond to the public key 

(Also, minisign no longer requires the password).

**Review by ErikSchierboom — APPROVED** [review]

**Comment by ee7** [comment]
I've created a draft release in this repo - see the edited demonstration in the first post.

**Review by ErikSchierboom — APPROVED** [review]

**Pull request merged**